contributors/design-proposals/api-paging.md

```diff
@@ -0,0 +1,148 @@
+# Reduce the amount of client and server memory allocated during large API lists
+
+On large clusters, performing API queries that return all of the objects of a given resource type (GET /api/v1/pods, GET /api/v1/secrets) can lead to significant variations in peak memory use on the server and contribute substantially to long tail request latency.
+
+When loading very large sets of objects -- some clusters are now reaching 100k pods or equivalent numbers of supporting resources -- the system must:
+
+* Construct the full range description in etcd in memory and serialize it as protobuf in the client
+  * Some clusters have reported over 500MB being stored in a single object type
```

Comment:
Quick question. One of the issues is the memory consumption at the server side, this paper calls for limits to be done at the client. I understand the reason is to be compatible, but that leaves the issue at the server still, if I am correct, because the client may choose not to send the `limit` query. If this is true, I would suggest a model which the server automatically limits the response. The servers know how large the payload would be and could predict how much memory it would consume, optimizing accordingly. I think leaving this to the client may solve only a part of the issue if the client decides to not send the `limit` query.

I do understand that we need to be backwards compatible, but I would suggest that the limit is _also_ done at the server including limit from the client. The server would then use the minimum limit value.

Reply:
Clients have to opt in.  If users want to fix the memory problem, they have to upgrade.  We can't break the semantics of the existing APIs for existing clients, so everything follows from that.